core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -257,22 +257,48 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
     val partitionsByDestination: immutable.Map[Node, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
       var brokerNode: Option[Node] = None
 
-      // TODO: instead of retry until succeed, we can first put it into an unknown broker queue and let the sender thread to look for its broker and migrate them
-      while (brokerNode.isEmpty) {
-        brokerNode = metadataCache.getPartitionLeaderEndpoint(topicPartition.topic, topicPartition.partition, interBrokerListenerName)
+      brokerNode = metadataCache.getPartitionLeaderEndpoint(topicPartition.topic, topicPartition.partition, interBrokerListenerName)
```

Comment:
1. good question about `getPartitionLeaderEndpoint`: I will use a separate queue for brokers who are not known but not available yet.
2. that is intentional: if one of the partitions are deleted but others have successfully written the markers, we should still treat this transaction ad completed, since the append of prepareXX txn log means that the txn has to be completedXX, even if some data partitions are deleted.